File: backend/api/schemas/analysis.py
"""Analysis request and response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class AnalysisRequest(BaseModel):
    scenario_selected: str = Field(default="normal.csv")
    event_mode: bool = Field(default=False)


class FeedbackRequest(BaseModel):
    verdict: str = Field(min_length=2, max_length=120)


class ScenarioSummary(BaseModel):
    slug: str
    label: str
    filename: str
    description: str
    occupancy_mode: str | None = None
    expected_has_leak: bool | None = None


class AnalysisResponse(BaseModel):
    analysis_id: str | None = None
    has_leak: bool
    leak_lpm: float
    total_liters: float
    leak_type: str | None = None
    confidence: float
    event_mode: bool
    event_rows: int
    source_mode: str
    scenario_selected: str | None = None
    validation_summary: dict[str, Any]
    reasoning_string: str
    financial_loss: dict[str, Any]
    environmental_impact: dict[str, Any]
    insights: dict[str, Any]
    anomalies: list[dict[str, Any]]
    telemetry_points: list[dict[str, Any]]


class AnalysisHistoryItem(BaseModel):
    analysis_id: str
    scenario_selected: str
    event_mode: bool
    has_leak: bool
    confidence: float
    leak_lpm: float
    total_liters: float
    created_at: str


class FeedbackResponse(BaseModel):
    analysis_id: str
    feedback: str
    confidence: float
    predicted_leak: bool

File: frontend/services/analysis.ts
import { apiFetch } from "@/lib/api";

export type ScenarioSummary = {
  slug: string;
  label: string;
  filename: string;
  description: string;
  occupancy_mode?: string;
  expected_has_leak?: boolean;
};

export type AnalysisRequest = {
  scenario_selected: string;
  event_mode: boolean;
};

export type AnalysisResponse = {
  analysis_id: string | null;
  has_leak: boolean;
  leak_lpm: number;
  total_liters: number;
  leak_type: string | null;
  confidence: number;
  event_mode: boolean;
  event_rows: number;
  source_mode: string;
  scenario_selected: string | null;
  validation_summary: Record<string, unknown>;
  reasoning_string: string;
  financial_loss: Record<string, unknown>;
  environmental_impact: Record<string, unknown>;
  insights: Record<string, unknown>;
  anomalies: Array<Record<string, unknown>>;
  telemetry_points: Array<Record<string, unknown>>;
};

export type AnalysisHistoryItem = {
  analysis_id: string;
  scenario_selected: string;
  event_mode: boolean;
  has_leak: boolean;
  confidence: number;
  leak_lpm: number;
  total_liters: number;
  created_at: string;
};

export type FeedbackResponse = {
  analysis_id: string;
  feedback: string;
  confidence: number;
  predicted_leak: boolean;
};

export function fetchScenarios() {
  return apiFetch<ScenarioSummary[]>("/scenarios");
}

export function runAnalysis(payload: AnalysisRequest) {
  return apiFetch<AnalysisResponse>("/analyses", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function fetchAnalysisHistory() {
  return apiFetch<AnalysisHistoryItem[]>("/analyses");
}

export function fetchAnalysisById(analysisId: string) {
  return apiFetch<AnalysisResponse>(`/analyses/${analysisId}`);
}

export function submitFeedback(analysisId: string, verdict: string) {
  return apiFetch<FeedbackResponse>(`/analyses/${analysisId}/feedback`, {
    method: "POST",
    body: JSON.stringify({ verdict }),
  });
}